### ISSUE TITLE
attempting to tokenize lots of open parens results in a stack overflow

### ISSUE THREAD

**Issue by frewsxcv** [alexcrichton/proc-macro2#55]
https://gist.github.com/frewsxcv/3b427948da01685c95c26ab4a005c337

```
> cargo run
   Compiling proc-macro2-foo v0.1.0 (file:///private/tmp/proc-macro2-foo)
    Finished dev [unoptimized + debuginfo] target(s) in 2.4 secs
     Running `target/debug/proc-macro2-foo`

thread 'main' has overflowed its stack
fatal runtime error: stack overflow
fish: 'cargo run' terminated by signal SIGABRT (Abort)
```

found via [afl.rs](https://github.com/rust-fuzz/afl.rs)

**Comment by dtolnay** [comment]
Looks like rustc has the same behavior as of 1.25.0-nightly.

```rust
fn main() {
    let s = ((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((...
}
```

```
thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by CAD97** [comment]
stable still overflows the stack, but beta no longer does unless you actually close all of the parenthesis. The example I constructed by "fixing" the given example has 23,512 open parenthesis.

https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=381e970dd9063ff8757a1a00a3573a40

**Comment by CAD97** [comment]
Correction: I had bad info somehow it still fails

**Comment by dtolnay** [comment]
Fixed by #231 + #232.
